### PR TITLE
validate nonResourceURL in create clusterrole

### PR DESCRIPTION
**Pull request by xilabao**
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 1]
```diff
@@ -133,6 +134,12 @@ func (c *CreateClusterRoleOptions) Validate() error {
 				return fmt.Errorf("invalid verb: '%s' for nonResourceURL", v)
 			}
 		}
+
+		for _, nonResourceURL := range c.NonResourceURLs {
+			if !strings.HasPrefix(nonResourceURL, "/") {
```
Comment (CaoShuFeng) [inline pkg/kubectl/cmd/create_clusterrole.go]: A single "*" is also allowed nonResourceURL.
But "*" is allowed in the tail, so this is invalid: /foo/*bar

**Review by CaoShuFeng — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 2]
```diff
@@ -133,6 +134,12 @@ func (c *CreateClusterRoleOptions) Validate() error {
 				return fmt.Errorf("invalid verb: '%s' for nonResourceURL", v)
 			}
 		}
+
+		for _, nonResourceURL := range c.NonResourceURLs {
+			if nonResourceURL != "*" && !strings.HasPrefix(nonResourceURL, "/") {
+				return fmt.Errorf("invalid nonResourceURL: '%s'", nonResourceURL)
```
Comment (liggitt) [inline pkg/kubectl/cmd/create_clusterrole.go]: make the message a little more helpful (like "nonResourceURL should start with /")

File: pkg/kubectl/cmd/create_clusterrole.go (hunk 1 above)
Comment (xilabao) [inline pkg/kubectl/cmd/create_clusterrole.go]: add validate of '*', does it neccesary to validate '/foo/*bar' ?
the server side doesn't validate the nonResourceURL, should we validate in the server side too? @liggitt

Comment (liggitt) [inline pkg/kubectl/cmd/create_clusterrole.go]: no server side validation, just make sure there's no more than one `*` here, and if there is one, it is at the end

**Comment by xilabao** [comment]
Done. PTAL

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 3]
```diff
@@ -133,6 +134,21 @@ func (c *CreateClusterRoleOptions) Validate() error {
...
+			if nonResourceURL == "*" {
+				continue
+			}
+
+			if !strings.HasPrefix(nonResourceURL, "/") {
+				return fmt.Errorf("nonResourceURL should start with /")
+			}
+
+			count := strings.Count(nonResourceURL, "*")
+			if count > 1 || (count == 1 && !strings.HasSuffix(nonResourceURL, "/*")) {
```
Comment (CaoShuFeng) [inline pkg/kubectl/cmd/create_clusterrole.go]: I think something like /logs/kube* is also allowed.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go#L123

Comment (xilabao) [inline pkg/kubectl/cmd/create_clusterrole.go]: fixed. thx

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 4]
```diff
@@ -133,6 +134,21 @@ func (c *CreateClusterRoleOptions) Validate() error {
...
+			if nonResourceURL == "*" {
+				continue
+			}
+
+			if !strings.HasPrefix(nonResourceURL, "/") {
+				return fmt.Errorf("nonResourceURL should start with /")
+			}
+
+			count := strings.Count(nonResourceURL, "*")
+			if count > 1 || (count == 1 && !strings.HasSuffix(nonResourceURL, "*")) {
```
Comment (CaoShuFeng) [inline pkg/kubectl/cmd/create_clusterrole.go]: nit: a better way of expressing line 147,148 would be
strings.ContainsRune(url[:len(url)-1], '*')

File: pkg/kubectl/cmd/create_clusterrole.go [hunk 5]
```diff
@@ -133,6 +134,20 @@ func (c *CreateClusterRoleOptions) Validate() error {
...
+			if nonResourceURL == "*" {
+				continue
+			}
+
+			if nonResourceURL == "" || !strings.HasPrefix(nonResourceURL, "/") {
+				return fmt.Errorf("nonResourceURL should start with /")
+			}
+
+			if strings.ContainsRune(nonResourceURL[:len(nonResourceURL)-1], '*') {
+				return fmt.Errorf("nonResourceURL not support more than one '*', and if there is one, it is at the end")
```
Comment (liggitt) [inline pkg/kubectl/cmd/create_clusterrole.go]: `nonResourceURL only supports wildcard matches when '*' is at the end`

**Comment by liggitt** [comment]
nit on text, lgtm otherwise

File: pkg/kubectl/cmd/create_clusterrole.go (hunk 5 above)
Comment (xilabao) [inline pkg/kubectl/cmd/create_clusterrole.go]: updated. @liggitt PTAL

**Pull request merged**